From c0f15e7e61fbd5c3a91a56fdcd8a2cf7508b8c41 Mon Sep 17 00:00:00 2001
From: Alexey Kodanev <aleksei.kodanev@bell-sw.com>
Date: Tue, 26 Dec 2023 10:27:22 +0000
Subject: [PATCH] database: verify old file onwers with fallback paths

This is because the old and new package versions may have different
dir-paths pointing to the same location (e.g. /lib -> usr/lib). In
this particular case, the previous owner won't be found when
upgrading and downgrading.

Fix it for system symlinks: lib, bin and sbin.
---
 src/database.c | 83 +++++++++++++++++++++++++++++++++++++++++++++++---
 1 file changed, 78 insertions(+), 5 deletions(-)

diff --git a/src/database.c b/src/database.c
index c742ef0..0f33ee1 100644
--- a/src/database.c
+++ b/src/database.c
@@ -2932,8 +2932,55 @@ static void apk_db_purge_pkg(struct apk_database *db, struct apk_installed_packa
 	}
 }
 
+static bool dir_incl_path(const apk_blob_t *dir, const apk_blob_t *path)
+{
+	if (dir->len < path->len || strncmp(dir->ptr, path->ptr, path->len))
+		return false;
+	if (dir->len > path->len && dir->ptr[path->len] != '/')
+		return false;
+
+	return true;
+}
+
+static apk_blob_t make_fallback_dir(struct apk_db_dir *dir, char *tmpdir)
+{
+	apk_blob_t orig = APK_BLOB_PTR_LEN(dir->name, dir->namelen);
+	apk_blob_t links[] = {
+		{ .ptr = "lib",  .len = 3},
+		{ .ptr = "bin",  .len = 3},
+		{ .ptr = "sbin", .len = 4},
+	};
+	apk_blob_t d = orig;
+	size_t i, remove_usr_path = 0;
+
+	if (d.len >= 4 && strncmp(d.ptr, "usr/", 4) == 0) {
+		remove_usr_path = 1;
+		d.ptr += 4;
+		d.len -= 4;
+	}
+	for (i = 0; i < ARRAY_SIZE(links); ++i) {
+		if (dir_incl_path(&d, &links[i]))
+			break;
+	}
+	if (i >= ARRAY_SIZE(links))
+		return orig;
+
+	if (remove_usr_path)
+		return d;
+
+	/* add 'usr/' + '\0' */
+	d.len += 4;
+	if (d.len + 1 > PATH_MAX)
+		return orig;
+	snprintf(tmpdir, d.len + 1, "usr/%s", d.ptr);
+	d.ptr = tmpdir;
+
+	return d;
+}
+
 static uint8_t apk_db_migrate_files_for_priority(struct apk_database *db,
 						 struct apk_installed_package *ipkg,
+						 struct apk_package *oldpkg,
 						 uint8_t priority)
 {
 	struct apk_out *out = &db->ctx->out;
@@ -2947,6 +2994,7 @@ static uint8_t apk_db_migrate_files_for_priority(struct apk_database *db,
 	apk_blob_t dirname;
 	int r, ctrl, inetc;
 	uint8_t dir_priority, next_priority = APK_FS_PRIO_MAX;
+	char tmpdir[PATH_MAX];
 
 	hlist_for_each_entry_safe(diri, dc, dn, &ipkg->owned_dirs, pkg_dirs_list) {
 		dir = diri->dir;
@@ -2974,6 +3022,32 @@ static uint8_t apk_db_migrate_files_for_priority(struct apk_database *db,
 			ofile = (struct apk_db_file *) apk_hash_get_hashed(
 				&db->installed.files, APK_BLOB_BUF(&key), hash);
 
+			/* re-check with fallback directory on upgrade/downgrade */
+			if (!ofile && oldpkg && dir->namelen >= 3) {
+				apk_blob_t new_dir = make_fallback_dir(dir, tmpdir);
+
+				if (new_dir.ptr != dir->name) {
+					if (out->verbosity >= 2) {
+						apk_msg(out, "re-check file %.*s (%.*s) in new dir: %.*s",
+							(int)file->namelen, file->name,
+							(int)dir->namelen, dir->name,
+							(int)new_dir.len, new_dir.ptr);
+					}
+
+					key = (struct apk_db_file_hash_key) {
+						.dirname = new_dir,
+						.filename = APK_BLOB_PTR_LEN(file->name, file->namelen),
+					};
+					unsigned long dir_hash = apk_hash_from_key(&db->installed.dirs, key.dirname);
+					hash = apk_blob_hash_seed(key.filename, dir_hash);
+
+					ofile = (struct apk_db_file *) apk_hash_get_hashed(
+						&db->installed.files, APK_BLOB_BUF(&key), hash);
+					if (ofile && out->verbosity >= 2)
+						apk_msg(out, "re-check: found previous file");
+				}
+			}
+
 			if (!file->broken) {
 				ctrl = APK_FS_CTRL_COMMIT;
 				if (ofile && ofile->diri->pkg->name == NULL) {
@@ -2998,7 +3072,6 @@ static uint8_t apk_db_migrate_files_for_priority(struct apk_database *db,
 						    db->ctx->apknew_suffix);
 					}
 				}
-
 				// Commit changes
 				r = apk_fsdir_file_control(&d, key.filename, ctrl);
 				if (r < 0) {
@@ -3018,7 +3091,6 @@ static uint8_t apk_db_migrate_files_for_priority(struct apk_database *db,
 					}
 				}
 			}
-
 			// Claim ownership of the file in db
 			if (ofile != file) {
 				if (ofile != NULL) {
@@ -3037,10 +3109,11 @@ static uint8_t apk_db_migrate_files_for_priority(struct apk_database *db,
 }
 
 static void apk_db_migrate_files(struct apk_database *db,
-				 struct apk_installed_package *ipkg)
+				 struct apk_installed_package *ipkg,
+				 struct apk_package *oldpkg)
 {
 	for (uint8_t prio = APK_FS_PRIO_DISK; prio != APK_FS_PRIO_MAX; )
-		prio = apk_db_migrate_files_for_priority(db, ipkg, prio);
+		prio = apk_db_migrate_files_for_priority(db, ipkg, oldpkg, prio);
 }
 
 static int apk_db_unpack_pkg(struct apk_database *db,
@@ -3156,7 +3229,7 @@ int apk_db_install_pkg(struct apk_database *db, struct apk_package *oldpkg,
 			apk_pkg_uninstall(db, newpkg);
 			goto ret_r;
 		}
-		apk_db_migrate_files(db, ipkg);
+		apk_db_migrate_files(db, ipkg, oldpkg);
 	}
 
 	if (oldpkg != NULL && oldpkg != newpkg && oldpkg->ipkg != NULL) {
-- 
2.25.1

